Sources/LanguageServerProtocolJSONRPC/JSONRPCConnection.swift

```diff
@@ -37,51 +190,50 @@ public final class JSONRPCConnection {
   /// Current state of the connection, used to ensure correct usage.
   var state: State
 
-  /// *Public for testing* Buffer of received bytes that haven't been parsed.
-  public var _requestBuffer: [UInt8] = []
-
   private var _nextRequestID: Int = 0
 
-  struct OutstandingRequest {
-    var requestType: _RequestType.Type
-    var responseType: ResponseType.Type
-    var queue: DispatchQueue
-    var replyHandler: (LSPResult<Any>) -> Void
-  }
-
   /// The set of currently outstanding outgoing requests along with information about how to decode and handle their responses.
   var outstandingRequests: [RequestID: OutstandingRequest] = [:]
 
   var closeHandler: (() -> Void)! = nil
 
+  /// Initialise a `JSONRPCConnection`.
+  ///
+  /// - parameters:
+  ///    - group: The `EventLoopGroup` on which to perform the I/O operations.
+  ///    - inputFileHandle: The `FileHandle` on which to receive JSON-RPC requests.
+  ///    - outputFileHandle: The `FileHandle` on which to send JSON-RPC responses.
+  ///    - takeFileDescriptorOwnership: If `true` thie `JSONRPCConnection` will duplicate the file descriptors and
+  ///                                   close the ones the were passed in as `inputFileHandle` and `outputFileHandle`.
+  ///                                   If in doubt, set this to `true` because it prevents the caller from accidentally
+  ///                                   closing file descriptors that are still in use.
+  ///    - syncRequests: *For Testing* Whether to wait for requests to finish before handling the next message.
+  ///    - closeHandler: Handler to execute when this `JSONRPCConnection` is closed.
   public init(
+    group: EventLoopGroup,
```

Comment:
Doing that in `_close` is dangerous because once the EventLoopGroup is shut down, nothing can run on them anymore. Ideally, there should be one EventLoopGroup and it should usually be global to the application (much like libdispatch's thread pool that is implicitly always there).

For some workloads (and tests) we have to give users exact control over the EventLoopGroup(s) which is why we don't just have a global one like Dispatch.

If you really want to make NIO an implementation detail, then I guess making a global one inside of `JSONRPCConnection.swift` (that never gets shut down, like Dispatch's) is the best solution.

From a testing PoV however it is much nicer to spawn the resources that you need at the beginning of the test  and shut them down at the end as it's done now. That does however mean wiring it through everywhere which I believe is good (explicit dependency injection) but I can see that not everybody will like it :).

Reply:
> Doing that in _close is dangerous because once the EventLoopGroup is shut down, nothing can run on them anymore.

If we're closing the connection, what would we still want to be able to run?

> Ideally, there should be one EventLoopGroup and it should usually be global to the application (much like libdispatch's thread pool that is implicitly always there).

Why would I want one per application instead of one per `JSONRPCConnection`? We are using `numberOfThreads: 1` since that's the model that makes sense for an individual `JSONRPCConnection`, but there's no intrinsic reason another connection should share that single thread.

Reply:
... or would the answer be that if we use a global eventloopgroup that we would increase the number of thread?

Reply:
@benlangmuir 

> Why would I want one per application instead of one per JSONRPCConnection?

NIO makes an effort to not burn one thread per connection by leveraging asynchronous IO. The result is that we can handle an arbitrarily high number of connections on just one thread. You _can_ use multiple threads if the pure CPU load of those connections actually maxes out one core.

> Why would I want one per application instead of one per JSONRPCConnection? 

Because all JSONRPCConnections in the application (and potentially other connections in the future) can share that EventLoopGroup.

> ... or would the answer be that if we use a global eventloopgroup that we would increase the number of threads?

No, I wouldn't expect sourcekit-lsp to ever need more than one event loop thread to do all its IO work.

Reply:
Okay, I'm convinced there should only be one :smile:  I still want NIO to be an implementation detail of JSONRPCConnection though. I can see why having deep control over it for testing might be cleaner, but it's exposing too much about the RPC to the rest of the code.  If it's just a question of having access in testing, I think we can expose the singleton EventLoopGroup just for the tests.  It's also a bit lopsided since the majority of our code uses Dispatch, and as you say the runtime for that is implicit and global. If we someday added an XPC or other non-NIO-based RPC mechanism, having an EventLoopGroup floating around that we don't use is a nuisance.

Reply:
@benlangmuir Sounds good! I wasn't 100% sure now how thoroughly you wanted to remove the explicit dependency injection and make it global/implicit. I've added a commit (which I'll squash in if you're happy with this) which achieves two things:

1. In non-test code there is still full control over the EventLoopGroup(s) used
2. In most tests, we don't actually have to know about NIO.

Reply:
> In non-test code there is still full control over the EventLoopGroup(s) used

Sorry should have been clearer: I don't want us to need the event loop group anywhere outside JSONRPCConnection (and possibly tests for the connection), so I want to remove the uses in BuildServerBuildSystem, SourceKitServer, etc.

Reply:
Cool, will finish that on Monday.

Reply:
@benlangmuir done, better now?